### PR TITLE
Implement `Type is Type`

### PR DESCRIPTION
**Pull request by ntrel**
This supplants #759. The same rationale applies for requiring `(identifier) is` to test an expression, see that PR for details. I added a diagnostic for `identifier is expr` where `expr` is not a type-id, because `Type is expr` is not valid - instead it will suggest using `(identifier) is`.

`cpp2util.h` changes are taken from #701 by @filipsajdak. That makes `std::runtime_error is std::exception` true, which seems correct, although I think that is not mentioned in P2392.

I had to add a `speculative` parameter to `parser::type_id()` which is `true` when there might not be a type (but an expression instead). This avoids an error for `ptr*` which is parsed as a postfix-expression from prefix-expression from is-as-expression.

### CODE REVIEW

File: include/cpp2util.h [hunk 1]
```diff
@@ -959,6 +959,15 @@ inline auto to_string(auto&& value, std::string_view) -> std::string
...
 using empty = void;
 
+//  Type is Type
+//
+
+template <typename X, typename C>
+auto is() -> std::false_type { return {}; }
+
+template <typename X, typename C>
+    requires std::same_as<X, C> || std::derived_from<X,C>
```
Comment (JohelEGP) [inline include/cpp2util.h:969]:
```suggestion
    requires std::derived_from<X,C>
```
`std::same_as<X, C> <= std::derived_from<X, C>` (implication).

File: source/parse.h [hunk 2]
```diff
@@ -6015,10 +6030,10 @@ class parser
 
     //G is-as-expression:
     //G     prefix-expression
+    //GTODO     type-id is-type-constraint
```
Comment (JohelEGP) [inline source/parse.h:6033]:
```suggestion
    //G     type-id is-type-constraint
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: include/cpp2util.h (hunk 1 above)
Comment (JohelEGP) [reply include/cpp2util.h:969]: Sorry, I was wrong.
`std::same_as` can be `true` for non-class types,
whereas `std::derived_from` can only be `true` for classes.

**Comment by JohelEGP** [comment]
What about the [_alternative_](https://github.com/hsutter/cppfront/issues/387#is-as-expression-target) grammar?

**Comment by JohelEGP** [comment]
> I added a diagnostic for `identifier is expr` where `expr` is not a type-id, because `Type is expr` is not valid - instead it will suggest using `(identifier) is`.

If this implements <https://github.com/hsutter/cppfront/issues/358#issuecomment-1516815580>,
please add "Closes #358" to the opening comment.

**Comment by hsutter** [comment]
Hi! Sorry it took me so long to get to this one. I'm going to close it as probably-dated now, but if you want to pursue it please reopen and refresh. Thanks again, and sorry again for the lag.

**Pull request closed without merging**